In [49]:
pip install streamlit pandas plotly


Defaulting to user installation because normal site-packages is not writeable
   ---------------------------------------- 0.0/9.2 MB ? eta -:--:--
   ------------------ --------------------- 4.2/9.2 MB 20.9 MB/s eta 0:00:01
   ---------------------------------------  9.2/9.2 MB 25.9 MB/s eta 0:00:01
   ---------------------------------------- 9.2/9.2 MB 22.9 MB/s  0:00:00
   ---------------------------------------- 0.0/797.5 kB ? eta -:--:--
   ---------------------------------------- 797.5/797.5 kB 16.7 MB/s  0:00:00
   ---------------------------------------- 0.0/11.3 MB ? eta -:--:--
   -------------------------- ------------- 7.6/11.3 MB 36.2 MB/s eta 0:00:01
   ---------------------------------------  11.3/11.3 MB 29.4 MB/s eta 0:00:01
   ---------------------------------------- 11.3/11.3 MB 25.3 MB/s  0:00:00
   ---------------------------------------- 0.0/9.9 MB ? eta -:--:--
   ------------- -------------------------- 3.4/9.9 MB 18.3 MB/s eta 0:00:01
   ---------------------- -


[notice] A new release of pip is available: 25.2 -> 26.1.2
[notice] To update, run: C:\Users\kumar\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [48]:
clean_df["category"].value_counts()

category
Other            632
App Bug          345
Delivery         285
Staff/Support    242
Billing          217
Name: count, dtype: int64

In [47]:
df["category"].value_counts()

KeyError: 'category'

In [46]:
clean_df["sentiment"] = clean_df["feedback_clean"].apply(get_sentiment)

clean_df["priority"] = clean_df.apply(
    get_priority,
    axis=1
)

In [45]:
negative_words.extend([
    "cannot",
    "can't",
    "did not",
    "not working",
    "failed",
    "greyed out",
    "stuck",
    "missing",
    "refund",
    "charged",
    "late",
    "cancelled"
])

In [44]:
df["category"].value_counts()

KeyError: 'category'

In [43]:
clean_df[
[
    "feedback_clean",
    "category",
    "sentiment",
    "priority",
    "issue_summary"
]
].head()

,feedback_clean,category,sentiment,priority,issue_summary
0,thank you to the agent priya who handled my co...,Staff/Support,Neutral,Low,thank you to the agent priya who handled my co...
1,"five stars, the app is fantastic and reliable ()",App Bug,Positive,Low,"five stars, the app is fantastic and reliable ()"
2,"cannot add address, the save button is greyed out",Other,Negative,High,"cannot add address, the save button is greyed out"
3,works fine for me most of the time,Other,Neutral,Low,works fine for me most of the time
4,coupon save50 did not apply at checkout and i ...,Billing,Negative,High,coupon save50 did not apply at checkout and i ...


In [42]:
clean_df.to_csv(
    "../output/enriched_feedback.csv",
    index=False
)

print("Enriched file exported successfully")

Enriched file exported successfully


In [41]:
clean_df.columns

Index(['id', 'timestamp', 'source', 'rating', 'feedback_text', 'text_length',
       'timestamp_clean', 'feedback_clean', 'category', 'sentiment',
       'priority', 'issue_summary'],
      dtype='object')

In [40]:
clean_df.to_csv(
    "../output/enriched_feedback.csv",
    index=False
)

print("Export completed")

Export completed


In [39]:
clean_df["priority"] = (
    clean_df.apply(
        get_priority,
        axis=1
    )
)

In [38]:
clean_df["sentiment"] = (
    clean_df["feedback_clean"]
    .apply(get_sentiment)
)


In [36]:
clean_df[
[
    "feedback_clean",
    "category",
    "sentiment",
    "priority",
    "issue_summary"
]
]

,feedback_clean,category,sentiment,priority,issue_summary
0,thank you to the agent priya who handled my co...,Staff/Support,Neutral,Low,thank you to the agent priya who handled my co...
1,"five stars, the app is fantastic and reliable ()",App Bug,Positive,Low,"five stars, the app is fantastic and reliable ()"
2,"cannot add address, the save button is greyed out",Other,Neutral,Low,"cannot add address, the save button is greyed out"
3,works fine for me most of the time,Other,Neutral,Low,works fine for me most of the time
4,coupon save50 did not apply at checkout and i ...,Billing,Neutral,Low,coupon save50 did not apply at checkout and i ...
...,...,...,...,...,...
1805,food was spilled all over the bag when it reac...,Other,Neutral,Low,food was spilled all over the bag when it reached
1806,the new version drains my battery so fast,Other,Positive,Low,the new version drains my battery so fast
1807,food was spilled all over the bag when it reac...,Staff/Support,Neutral,Low,food was spilled all over the bag when it reached
1808,custooomer care kept me on hold for 40 minutes...,Other,Neutral,Low,custooomer care kept me on hold for 40 minutes...


In [34]:
def summarize_issue(text):

    text = str(text)

    words = text.split()

    return " ".join(words[:10])

clean_df["issue_summary"] = (
    clean_df["feedback_clean"]
    .apply(summarize_issue)
)

In [33]:
clean_df["priority"].value_counts()

priority
Low       1194
High       305
Medium     222
Name: count, dtype: int64

In [32]:
def get_priority(row):

    rating = row["rating"]

    sentiment = row["sentiment"]

    if sentiment == "Negative":

        if pd.notna(rating):

            if rating <= 2:
                return "High"

            elif rating <= 3:
                return "Medium"

        return "Medium"

    return "Low"

clean_df["priority"] = (
    clean_df.apply(
        get_priority,
        axis=1
    )
)

In [31]:
clean_df["sentiment"].value_counts()

sentiment
Neutral     915
Negative    527
Positive    279
Name: count, dtype: int64

In [37]:
positive_words = [
    "great", "excellent", "good", "amazing",
    "love", "fantastic", "smooth", "fast",
    "helpful", "easy", "perfect"
]

negative_words = [
    "bad",
    "worst",
    "late",
    "delay",
    "crash",
    "error",
    "refund",
    "charged",
    "missing",
    "broken",
    "rude",
    "issue",
    "problem",
    "failed",
    "cancelled",
    "cannot",
    "cant",
    "did not",
    "not working",
    "stuck",
    "greyed out",
    "greyed out",
    "slow",
    "lost",
    "wrong",
    "never arrived",
    "not delivered"
]

def get_sentiment(text):

    text = str(text).lower()

    pos = sum(word in text for word in positive_words)
    neg = sum(word in text for word in negative_words)

    if pos > neg:
        return "Positive"

    elif neg > pos:
        return "Negative"

    return "Neutral"

clean_df["sentiment"] = (
    clean_df["feedback_clean"]
    .apply(get_sentiment)
)

In [29]:
clean_df["category"].value_counts()

category
Other            632
App Bug          345
Delivery         285
Staff/Support    242
Billing          217
Name: count, dtype: int64

In [28]:
clean_df[
    ["feedback_clean","category"]
].sample(20)

,feedback_clean,category
1103,app stuck on loading screen for 5 minutes then...,App Bug
1468,ok i guess,Other
593,"love this app, ordering is so smooth and fast ()",App Bug
360,login button does nothing on the latest update...,App Bug
728,the map screen freezes and i cannot place an o...,App Bug
1383,the store staff was friendly when i picked up ...,Staff/Support
225,"perfect, the app froze again, what a wonderful...",App Bug
1681,please add more vegan restaurants to the platform,Other
81,what are your office timings on weekends,Other
129,chat support gave me copy paste answers that d...,Staff/Support


In [27]:
def classify_category(text):

    text = str(text).lower()

    if any(word in text for word in [
        "refund","charged","payment",
        "coupon","billing","subscription"
    ]):
        return "Billing"

    elif any(word in text for word in [
        "crash","bug","login","error",
        "freeze","app"
    ]):
        return "App Bug"

    elif any(word in text for word in [
        "delivery","driver","package",
        "late","missing item"
    ]):
        return "Delivery"

    elif any(word in text for word in [
        "support","agent","staff",
        "customer service"
    ]):
        return "Staff/Support"

    else:
        return "Other"

clean_df["category"] = clean_df["feedback_clean"].apply(classify_category)

In [26]:
clean_df.to_csv(
    "../output/cleaned_feedback.csv",
    index=False
)

print("Cleaned file exported")

Cleaned file exported


In [25]:
print(len(clean_df))

1721


In [24]:
def is_spam(text):
    text = str(text).strip().lower()

    if len(set(text)) <= 2 and len(text) > 5:
        return True

    return False

clean_df = clean_df[
    ~clean_df["feedback_clean"].apply(is_spam)
]

In [23]:
clean_df[
    ["feedback_text","feedback_clean"]
].sample(10)

,feedback_text,feedback_clean
1139,"Super easy to use and quick checkout, well don...","super easy to use and quick checkout, well don..."
608,"Package was left at the wrong door, I never re...","package was left at the wrong door, i never re..."
1191,"Great experience overall, food always fresh - ...","great experience overall, food always fresh - ..."
572,Got charged for a subscription I cancelled las...,got charged for a subscription i cancelled las...
1238,aaaaaaaaaaaaaaa,aaaaaaaaaaaaaaa
972,"Package was left at the wrong door, I never re...","package was left at the wrong door, i never re..."
518,Driver could not find my building and just can...,driver could not find my building and just can...
1491,FOOD WAS SPILLED ALL OVER THE BAG WHEN IT REAC...,food was spilled all over the bag when it reac...
556,"Random question, who designed your logo it loo...","random question, who designed your logo it loo..."
784,The store staff was friendly when I picked up ...,the store staff was friendly when i picked up ...


In [22]:
import re

def clean_text(text):
    text = str(text)

    text = text.lower()

    text = re.sub(r'\s+', ' ', text)

    text = re.sub(r'order\s*#?\d+', '', text)

    return text.strip()

clean_df["feedback_clean"] = (
    clean_df["feedback_text"]
    .apply(clean_text)
)

In [21]:
clean_df.sample(10)[
    [
        "feedback_text",
        "timestamp",
        "timestamp_clean"
    ]
]

,feedback_text,timestamp,timestamp_clean
1703,Thank you to the agent Priya who handled my co...,"March 11, 2024",2024-03-11
835,"Great experience overall, food always fresh Ag...",Jan 15 2024,2024-01-15
130,"My refund still hasn't come after 10 days, thi...",NaN,<NA>
0,THANK YOU TO THE AGENT PRIYA WHO HANDLED MY CO...,NaN,<NA>
135,Search bar shows no results even for items tha...,12-Jan-24,2024-01-12
1144,Order marked delivered but nothing arrived at ...,2024-01-22,2024-01-22
946,Delivery person was rude and threw the bag on ...,27/01/2024,2024-01-27
233,"Super easy to use and quick checkout, well don...","January 25, 2024",2024-01-25
698,"Average experience, could be better",NaN,<NA>
151,"App bahut acchi hai, bahut fast hai",2024-01-01 20:36:10,2024-01-01


In [20]:
print(clean_df.columns)


Index(['id', 'timestamp', 'source', 'rating', 'feedback_text', 'text_length',
       'timestamp_clean'],
      dtype='object')


In [19]:
clean_df.sample(10)[
    [
        "feedback_text",
        "feedback_clean",
        "timestamp",
        "timestamp_clean"
    ]
]

KeyError: "['feedback_clean'] not in index"

In [18]:
clean_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1725 entries, 0 to 1809
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   id               1725 non-null   int64  
 1   timestamp        1512 non-null   object 
 2   source           1725 non-null   object 
 3   rating           1336 non-null   float64
 4   feedback_text    1725 non-null   object 
 5   text_length      1725 non-null   int64  
 6   timestamp_clean  1512 non-null   object 
dtypes: float64(1), int64(2), object(4)
memory usage: 107.8+ KB


In [17]:
clean_df["timestamp_clean"].isna().sum()

213

In [16]:
clean_df[
    ["timestamp", "timestamp_clean"]
].head(20)

,timestamp,timestamp_clean
0,NaN,<NA>
1,02-Feb-24,2024-02-02
2,02/14/2024,2024-02-14
3,03/15/2024,2024-03-15
4,04/01/2024,2024-04-01
5,2024-03-25 16:31:48,2024-03-25
6,"March 18, 2024",2024-03-18
7,"January 21, 2024",2024-01-21
8,NaN,<NA>
9,13-Mar-24,2024-03-13


In [15]:
from dateutil import parser
import pandas as pd

def parse_date(date_value):
    try:
        return parser.parse(
            str(date_value),
            dayfirst=False
        ).strftime("%Y-%m-%d")
    except:
        return pd.NA

clean_df["timestamp_clean"] = (
    clean_df["timestamp"]
    .apply(parse_date)
)

Hidden Trap #1 Successfully Found

The dataset contains:

👍
meh
MEH
\t
nan

These are not customer complaints.
Removing them is correct and defensible.

------------------------------------------------------

Hidden Trap #2 Successfully Found
Date formats are mixed:

2024-02-24 21:38:08
02/28/2024
11/02/2024
Mar 14 2024
18-Feb-24
March 15, 2024
25/03/2024
2024-03-16 13:59:53
05-Feb-24

This is exactly the type of messy production data Bi3 expects you to handle.

In [14]:
clean_df["timestamp"].dropna().sample(30)

741     2024-02-24 21:38:08
803              02/28/2024
39               11/02/2024
748             Mar 14 2024
886              03/07/2024
528              03/14/2024
738              02/05/2024
1359            Mar 22 2024
562              2024-01-16
638              Jan 7 2024
1103             2024-02-07
794              2024-03-07
292               18-Feb-24
493          March 15, 2024
58               25/03/2024
667              Mar 6 2024
984              02/01/2024
369              02/18/2024
942              01/17/2024
473              Feb 3 2024
1763             02/25/2024
81              Mar 22 2024
193              13/02/2024
65           March 11, 2024
1246             02/03/2024
60               Jan 6 2024
902               18-Jan-24
923               19-Feb-24
354     2024-03-16 13:59:53
1600              05-Feb-24
Name: timestamp, dtype: object

In [13]:
print("Original Rows :", len(df))
print("Cleaned Rows :", len(clean_df))
print("Removed Rows :", len(df) - len(clean_df))


Original Rows : 1810
Cleaned Rows : 1725
Removed Rows : 85


In [12]:
clean_df = clean_df[
    clean_df["feedback_text"]
    .str.len() >= 5
]

In [11]:
meaningless_words = [
    "meh",
    "ok",
    "okay",
    "hmm",
    "nan",
    "n/a",
    "na",
    "👍"
]

clean_df = clean_df[
    ~clean_df["feedback_text"]
    .str.lower()
    .isin(meaningless_words)
]


In [10]:
clean_df["feedback_text"] = (
    clean_df["feedback_text"]
    .astype(str)
    .str.strip()
)

In [9]:
clean_df = clean_df[
    clean_df["feedback_text"]
    .astype(str)
    .str.strip() != ""
]


In [8]:
clean_df = clean_df[
    clean_df["feedback_text"].notna()
]

In [7]:
import pandas as pd
import numpy as np
import re

clean_df = df.copy()

# Phase 2 - Cleaning Strategy

The raw dataset contains several quality issues:

1. Missing feedback records
2. Missing timestamps
3. Inconsistent timestamp formats
4. Non-actionable feedback (e.g. "meh", emojis)
5. Placeholder values

Cleaning decisions:

- Remove rows without usable feedback
- Retain rows with missing timestamps
- Standardize timestamps to YYYY-MM-DD
- Remove non-actionable feedback
- Preserve legitimate repeated complaints

In [6]:
print("Shortest feedback messages")

df["feedback_text"] = df["feedback_text"].astype(str)

df["text_length"] = df["feedback_text"].str.len()

df.sort_values("text_length")[[
    "feedback_text",
    "text_length"
]].head(30)

Shortest feedback messages


,feedback_text,text_length
1130,👍,1
855,👍,1
1693,\t,1
35,👍,1
1339,\t,1
684,👍,1
1000,meh,3
1496,meh,3
906,meh,3
1607,meh,3


In [5]:
print("="*50)
print("EXACT DUPLICATES")
print("="*50)

print(df.duplicated().sum())

print("\n")

print("="*50)
print("SOURCE DISTRIBUTION")
print("="*50)

print(df["source"].value_counts())

print("\n")

print("="*50)
print("UNIQUE SOURCES")
print("="*50)

print(df["source"].unique())

print("\n")

print("="*50)
print("RATING DISTRIBUTION")
print("="*50)

print(df["rating"].value_counts(dropna=False))


EXACT DUPLICATES
0


SOURCE DISTRIBUTION
source
app_store_review    666
support_ticket      573
survey_comment      571
Name: count, dtype: int64


UNIQUE SOURCES
['app_store_review' 'survey_comment' 'support_ticket']


RATING DISTRIBUTION
rating
NaN    439
1.0    394
2.0    386
3.0    332
5.0    196
4.0     63
Name: count, dtype: int64


# Phase 1 - Data Audit

## Dataset Overview

Total Rows: 1810

Total Columns: 5

## Initial Observations

- Missing timestamps observed
- Multiple timestamp formats detected
- Missing ratings observed
- Text field contains customer feedback
- Source field contains 3 channels

## Next Step

Investigate duplicates and data quality issues.

In [4]:
print("="*50)
print("DATASET INFORMATION")
print("="*50)

df.info()

print("\n")
print("="*50)
print("MISSING VALUES")
print("="*50)

print(df.isnull().sum())

print("\n")
print("="*50)
print("EXACT DUPLICATES")
print("="*50)

print(df.duplicated().sum())

print("\n")
print("="*50)
print("SOURCE DISTRIBUTION")
print("="*50)

print(df["source"].value_counts())

print("\n")
print("="*50)
print("RATING DISTRIBUTION")
print("="*50)

print(df["rating"].value_counts(dropna=False))

DATASET INFORMATION
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1810 entries, 0 to 1809
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             1810 non-null   int64  
 1   timestamp      1587 non-null   object 
 2   source         1810 non-null   object 
 3   rating         1371 non-null   float64
 4   feedback_text  1800 non-null   object 
dtypes: float64(1), int64(1), object(3)
memory usage: 70.8+ KB


MISSING VALUES
id                 0
timestamp        223
source             0
rating           439
feedback_text     10
dtype: int64


EXACT DUPLICATES
0


SOURCE DISTRIBUTION
source
app_store_review    666
support_ticket      573
survey_comment      571
Name: count, dtype: int64


RATING DISTRIBUTION
rating
NaN    439
1.0    394
2.0    386
3.0    332
5.0    196
4.0     63
Name: count, dtype: int64


In [3]:
import pandas as pd
import numpy as np

df = pd.read_csv("../customer_feedback_raw 1.csv")

print("Shape:", df.shape)
df.head()

Shape: (1810, 5)


,id,timestamp,source,rating,feedback_text
0,1847,NaN,app_store_review,1.0,THANK YOU TO THE AGENT PRIYA WHO HANDLED MY CO...
1,1695,02-Feb-24,app_store_review,NaN,"Five stars, the app is fantastic and reliable ..."
2,2771,02/14/2024,app_store_review,1.0,"Cannot add address, the save button is greyed out"
3,2770,03/15/2024,survey_comment,3.0,Works fine for me most of the time
4,2075,04/01/2024,app_store_review,1.0,Coupon SAVE50 did not apply at checkout and I ...
